examples/xgboost/custom/xgboost_bagging_aggregator.py

```diff
@@ -0,0 +1,183 @@
+
+        The trees are pre-weighted during training.
+        Bagging aggregation simply add the new trees to existing global model.
+
+        Args:
+            expected_data_kind (Union[DataKind, Dict[str, DataKind]]):
+                DataKind for DXO. Defaults to DataKind.WEIGHTS
+                Indicating the tree representation given by XGBoost
+        """
+        super().__init__()
+        self.logger.debug(f"expected data kind: {expected_data_kind}")
+        self.history = []
+        self.expected_data_kind = expected_data_kind
+        # Check expected data kind
+        if expected_data_kind != DataKind.WEIGHTS:
+            raise ValueError(f"expected_data_kind = {expected_data_kind} is not {DataKind.WEIGHTS}")
+
+    def accept(self, shareable: Shareable, fl_ctx: FLContext) -> bool:
+        """Store shareable and update aggregator's internal state
+
+        Args:
+            shareable: information from contributor
+            fl_ctx: context provided by workflow
+
+        Returns:
+            The first boolean indicates if this shareable is accepted.
+            The second boolean indicates if aggregate can be called.
+        """
+        try:
+            dxo = from_shareable(shareable)
+        except BaseException:
+            self.log_exception(fl_ctx, "shareable data is not a valid DXO")
+            return False
+
+        contributor_name = shareable.get_peer_prop(key=ReservedKey.IDENTITY_NAME, default="?")
+        contribution_round = shareable.get_header(AppConstants.CONTRIBUTION_ROUND)
+
+        rc = shareable.get_return_code()
+        if rc and rc != ReturnCode.OK:
+            self.log_warning(fl_ctx, f"Contributor {contributor_name} returned rc: {rc}. Disregarding contribution.")
+            return False
+
+        # Accept expected DXO in shareable
+        if not isinstance(dxo, DXO):
+            self.log_error(fl_ctx, f"Expected DXO but got {type(dxo)}")
+            return False
+
+        if dxo.data_kind != self.expected_data_kind:
+            self.log_error(fl_ctx, "expected {} but got {}".format(self.expected_data_kind, dxo.data_kind))
+            return False
+
+        current_round = fl_ctx.get_prop(AppConstants.CURRENT_ROUND)
+        self.log_debug(fl_ctx, f"current_round: {current_round}")
+        if contribution_round != current_round:
+            self.log_warning(
+                fl_ctx,
+                f"discarding DXO from {contributor_name} at round: "
+                f"{contribution_round}. Current round is: {current_round}",
+            )
+            return False
+
+        for item in self.history:
+            if contributor_name == item["contributor_name"]:
+                prev_round = item["round"]
+                self.log_warning(
+                    fl_ctx,
+                    f"discarding DXO from {contributor_name} at round: "
+                    f"{contribution_round} as {prev_round} accepted already",
+                )
+                return False
```

Comment:
when will this situation happen?

Reply:
this indeed may not happen, but it is part of our current "dxo_aggregator", guess it is "in case"?